Implement hasFeaturesFrom helper

mihails-strasuns-sociomantic opened the pull request:
Follow-up to recent discussion about exposing library versions in D code for
`static if` shenanigans in downstream libraries.

leandro-lucarella-sociomantic commented on src/ocean/LibVersion.d:
```diff
@@ -0,0 +1,9 @@
+module ocean.LibVersion;
```
Since this module is starting to only care about new features introduced in minor versions, I wonder if it shouldn't be named `libFeatures` instead.

mihails-strasuns-sociomantic replied:
Sounds good.

leandro-lucarella-sociomantic reviewed (commented):
Besides the small bikeshedding, LGTM (not approving because it lacks release notes / docs, which I guess is intended being a WIP and all :P).

leandro-lucarella-sociomantic commented on src/ocean/LibVersion.d:
```diff
@@ -0,0 +1,9 @@
+module ocean.LibVersion;
+
+const has_features_2_6 = true;
```
Wait! Since previous versions won't have the `hasFeaturesFrom` (or even the features definition constants), does it make sense to define the features for old versions?

mihails-strasuns-sociomantic replied:
Yes, that is the plan. If design looks OK, I'll make separate pull requests per each supported branch to define consts separately.

mihails-strasuns-sociomantic replied:
See https://github.com/sociomantic-tsunami/ocean/pull/64

leandro-lucarella-sociomantic replied:
I still don't get it. Also, technically, this is a new feature (adding a module), so we shouldn't release that as a point release. We could make an exception, but I'd prefer not to unless there is rally a **very** strong case for it.

mihails-strasuns-sociomantic replied:
AFAICS it is neither a feature nor a bug fix - it is purely infrastructural maintenance for a feature that will be added in v2.7 (this PR). To make `hasFeatureFrom` helper reliable we need to add that module to all supported branches in all core libraries - it has no effect on their own API/functionality.

leandro-lucarella-sociomantic replied:
But again, you can't rely on this feature because then if you introduce this in v2.5.7, anyone still at v2.5.6 or lower will have the check failing. I know people is supposed to update to the latest patch release, but it won't always happen, and if you can't rely on this feature `const` being correctly defined, it means it's not really necessary to add the check at this point, so why add it?

For me is best and cleaner to do a clean cut and support this idiom only for v2.7+

mihails-strasuns-sociomantic replied:
> anyone still at v2.5.6 or lower will have the check failing

We don't support non-latest point releases, such users are strictly obliged to upgrade.

mihails-strasuns-sociomantic replied:
(same concern applies to v2.5.x and v2.6.x branches btw)

mathias-lang-sociomantic commented on src/ocean/core/VersionCheck.d:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    mixin("return is(typeof(Library.has_features_"
+        ~ ctfe_i2a(major) ~ "_" ~ ctfe_i2a(minor) ~ "));");
```
So in this case, how do we deprecate `has_features` symbols when a library is using `-de` ?

mihails-strasuns-sociomantic replied:
Same as usual. Library usage of `-de` does not propagate to applications - deprecating symbol will ensure it is not directly used in libraries but will keep working if app upgrades to ocean version above supported to library. Which is exactly what we want, right?

mathias-lang-sociomantic replied:
Oh I understand better the original discussion then. I like it :+1:

mihails-strasuns-sociomantic committed:
Implement hasFeaturesFrom helper

Intended to help with implementing optional features in downstream libraries.
Also provides feature constants for all version before 2.7

mihails-strasuns-sociomantic commented:
Update the PR

mihails-strasuns-sociomantic commented:
Green

leandro-lucarella-sociomantic commented on src/ocean/core/VersionCheck.d:
```diff
@@ -0,0 +1,49 @@
+    static assert ( hasFeaturesFrom!("ocean", 2, 0));
+    static assert ( hasFeaturesFrom!("ocean", 2, 6));
+    static assert ( hasFeaturesFrom!("ocean", 2, 7));
+    static assert (!hasFeaturesFrom!("ocean", 3, 0));
```
This unit test will have to be adapted when moving to v3. Why not just using a feature that will never be set. Otherwise are we updating this test each time we branch a new major? :)

mihails-strasuns-sociomantic replied:
Yeah makes sense

leandro-lucarella-sociomantic reviewed (approved):
Other than that (minor), LGTM.

daniel-zullo commented on src/ocean/LibFeatures.d:
```diff
@@ -0,0 +1,26 @@
+    later ocean versions.
+
+    Copyright:
+        Copyright (c) 2009-2016 Sociomantic Labs GmbH.
```
This is not right, we are in 2017.

mihails-strasuns-sociomantic replied:
https://github.com/sociomantic-tsunami/ocean/pull/66

daniel-zullo commented on src/ocean/core/VersionCheck.d:
```diff
@@ -0,0 +1,49 @@
+    module allows to query supported features with a standard API.
+
+    Copyright:
+        Copyright (c) 2009-2016 Sociomantic Labs GmbH.
```
2017

daniel-zullo commented on src/ocean/core/VersionCheck.d:
```diff
@@ -0,0 +1,49 @@
+                                                        // updated one day
+}
+
+private bool hasFeaturesFromFunc ( istring libname, ulong major, ulong minor ) ()
```
the methods in this module do not have documentation.

mihails-strasuns-sociomantic replied:
You are wrong on this part though :) The `ditto` bit will inherit public method documentation from the module header (and private one is not intended to be documented).

daniel-zullo replied:
Sorry but I don't totally agree as the parameters documentation would be missing if we use `ditto` to inherit from the module header.

gavin-norman-sociomantic replied:
I agree. This template and function ought to be documented in the usual style.

mathias-baumann-sociomantic replied:
This is still pending

daniel-zullo reviewed (commented):
Please be more careful before merging PRs without the right documentation.

mihails-strasuns-sociomantic commented:
Copyright statements are contagious disease that was artifically create by evil legal world to drive developers to madness :grumpy:
